src/main/java/present/Location.java

```diff
@@ -38,8 +46,100 @@ public double longitude() {
   public double distanceTo(Location other) {
     distanceComputations.incrementAndGet();
 
-    throw new UnsupportedOperationException("Please implement!");
+    return getDistanceFromCache(other);
   }
 
-  static AtomicInteger distanceComputations = new AtomicInteger();
+  /**
+   * Reads distance from cache. If found, returns it or else computes the distance, puts it in cache and then return it.
+   *
+   * @param other location to get the distance to
+   * @return distance to other location
+   */
+  private Double getDistanceFromCache(Location other) {
+    DistanceCacheKey distanceCacheKey = new DistanceCacheKey(this, other);
+    Double distance = DISTANCE_CACHE.get(distanceCacheKey);
+
+    if (distance != null) {
+      return distance;
+    }
+
+    distance = getDistance(other);
+
+    DISTANCE_CACHE.put(distanceCacheKey, distance);
+
+    return distance;
+  }
+
+  /**
+   * getDistance computes the distance from current location using the cos formula.
+   *
+   * @param other location to compute the distance to
+   * @return distance to other location
+   */
+  private Double getDistance (Location other) {
+    double currentLatitude = Math.toRadians(latitude);
+    double currentLongitude = Math.toRadians(longitude);
+    double otherLatitude = Math.toRadians(other.latitude);
+    double otherLongitude = Math.toRadians(other.longitude);
+
+    // Using the cos formula
+    double angleInRadians = Math.acos(
+            Math.sin(currentLatitude) * Math.sin(otherLatitude) +
+                    Math.cos(currentLatitude) * Math.cos(otherLatitude) * Math.cos(currentLongitude - otherLongitude)
+    );
+
+    return angleInRadians * ONE_RADIAN_IN_KM;
+  }
+
+  /**
+   * Private inner class for the cache key for DISTANCE_CACHE
+   *
+   * The motivation of this class is because distance between Point 1 to 2 is the same as Point 2 to 1. So this overrides equals to handle that use-case
+   */
+  private class DistanceCacheKey {
+    Location current;
+    Location other;
+
+    /**
+     * Construct a new DistanceCacheKey
+     *
+     * @param current location
+     * @param other location
+     */
+    public DistanceCacheKey(Location current, Location other) {
+      this.current = current;
+      this.other = other;
+    }
+
+    @Override
+    public boolean equals(Object object) {
+      if (object == null) {
+        return false;
+      }
+
+      if (object == this) {
+        return true;
+      }
+
+      if (!(object instanceof DistanceCacheKey)) {
+        return false;
+      }
+
+      DistanceCacheKey distanceCacheKey = (DistanceCacheKey) object;
+
+      return (current.equals(distanceCacheKey.current) && other.equals(distanceCacheKey.other)) ||
+              (current.equals(distanceCacheKey.other) && other.equals(distanceCacheKey.current));
+    }
+
+    @Override
+    public int hashCode() {
```

Comment:
The prime isn't necessary here. We could use:
```
public int hashCode() {
  return Objects.hashCode(current) + Objects.hashCode(other);
}
```
Even better, we could defend against nulls in the constructor and have:
```
public int hashCode() {
  return current.hashCode() + other.hashCode();
}
```

Reply:
Thank you for that. Let me know if you want me to change it.